pkg/buildkit/worker/leasemanager.go

```diff
@@ -0,0 +1,345 @@
+	}
+
+	m.log.Info("Scaling cluster, no unleased pods found")
+	if err = m.scaler.Scale(ctx, 1); err != nil {
+		return nil, err
+	}
+
+	m.log.Info("No unleased pods found, scaling workload")
+	if pod, err = m.getUnleasedPod(ctx); err != nil {
+		return nil, fmt.Errorf("unable to acquire lease after scaling: %w", err)
+	}
+
+	return pod, nil
+}
+
+func (m *podLeaseManager) getUnleasedPod(ctx context.Context) (*corev1.Pod, error) {
+	m.log.Info("Querying for pods", "namespace", m.namespace, "opts", m.podListOptions)
+	list, err := m.podClient.List(ctx, m.podListOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, pod := range list.Items {
+		if _, ok := pod.Annotations[leasedAnnotation]; !ok {
+			m.log.Info("Found unleased pod", "name", pod.Name, "namespace", pod.Namespace)
+			return &pod, nil
+		}
+	}
+
+	m.log.Info("No unleased pods found")
+	return nil, errNoUnleasedPods
+}
+
+func (m *podLeaseManager) buildURL(ctx context.Context, podName string) (string, error) {
+	m.log.Info("Querying for endpoints", "name", m.serviceName, "namespace", m.namespace)
+	endpoints, err := m.endpointsClient.Get(ctx, m.serviceName, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	var host string
+
+addressScan:
+	for _, subset := range endpoints.Subsets {
+		for _, address := range subset.Addresses {
+			if address.TargetRef.Name == podName {
+				for _, port := range subset.Ports {
+					if port.Port == m.servicePort {
+						host = strings.Join([]string{address.Hostname, endpoints.Name, endpoints.Namespace}, ".")
+						m.log.Info("Found eligible endpoint address", "host", host)
+
+						break addressScan
+					}
+				}
+			}
+		}
+	}
+
+	if host == "" {
+		return "", fmt.Errorf("endpoints %q does not expose pod %s on port %d", endpoints.Name, podName, 1234)
+	}
+
+	u, err := url.ParseRequestURI(fmt.Sprintf("tcp://%s:%d", host, m.servicePort))
+	if err != nil {
+		return "", fmt.Errorf("failed to parse endpoint url: %w", err)
+	}
+
+	return u.String(), nil
+}
+
+func (m *podLeaseManager) monitorPods(ctx context.Context) {
+	m.log.Info("Starting worker pod monitor", "syncTime", m.podSyncTime.String())
+	ticker := time.NewTicker(m.podSyncTime)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			m.log.Info("Shutting down worker pod monitor")
+			return
+		case <-ticker.C:
+			m.log.Info("Checking worker pool for expired leases")
+			if err := m.terminateUnleasedPods(ctx); err != nil {
+				m.log.Error(err, "Failed to clean worker pool")
+			}
+			m.log.Info("Worker pool lease check complete")
+		}
+	}
+}
+
+func (m *podLeaseManager) terminateUnleasedPods(ctx context.Context) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.log.Info("Querying for pods", "namespace", m.namespace, "opts", m.podListOptions)
+	list, err := m.podClient.List(ctx, m.podListOptions)
+	if err != nil {
+		return err
+	}
+
+	var removals []string
+	for idx := len(list.Items) - 1; idx >= 0; idx-- {
```

Comment:
Are the pods sorted ordinally?

Reply:
I did some testing with non-buildkit pods (different pod ages, names, workload types) and every invocation of the `List` func produced a slice of items sorted by name in ascending order. Yet, neither the [API docs](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.24/#list-pod-v1-core) nor the client-go source code indicate a "clear and intentional" sorting order. I did find a reddit post that says

>Someone at kubernetes slack channel answered my question:
>
>api server retrieves resources data from etcd and uses no sorting by default, which defaults to etcd default sorting, which is alphabetical

but that's hardly a credible source.

Despite the fact that the statefulset creates the pods in ascending order, I don't think there's any informantion to rely on this sorting. Furthermore, I can reverse sort this list and then range over it w/o having to use countdown `for` loop.